### PR TITLE
AST: Rename TypeRefinementContext to AvailabilityScope

### PR DESCRIPTION
**Pull request by tshortli**
Clarify the role `TypeRefinementContext` by renaming it to `AvailabilityScope`. There are a few reasons for doing this, despite the churn:

- `TypeRefinementContext` is a fairly broad name that doesn't provide much of a hint about the primary purpose of this structure for most contributors. Having `Availability` in the name is clarifying.
- This better aligns the names of files related to availability checking for code owner rules and file navigation.
- This infrastructure is already being heavily refactored, so it's a relatively good time to rename it.
- `AvailabilityScope` is evocative of `ASTScope`, which is useful because the two represent very similar concepts and have similar structures.
- The name "context" is overused in the compiler. Using "scope" instead allows us to use more distinct names in local variables, for example, which can clarify the code.

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please smoke test

File: include/swift/Option/Options.td [hunk 1]
```diff
@@ -1336,9 +1336,9 @@ def dump_scope_maps : Separate<["-"], "dump-scope-maps">,
   MetaVarName<"<expanded-or-list-of-line:column>">,
   ModeOpt,
   Flags<[FrontendOption, NoInteractiveOption, DoesNotAffectIncrementalBuild]>;
-def dump_type_refinement_contexts :
```
Comment (artemcm) [inline include/swift/Option/Options.td:1339]: Please also update `Options.swift` in [swift-driver](https://github.com/swiftlang/swift-driver).

**Review by artemcm — APPROVED** [review]
Cool!

**Comment by tshortli** [comment]
@swift-ci please smoke test

File: include/swift/Option/Options.td (hunk 1 above)
Comment (tshortli) [reply include/swift/Option/Options.td:1339]: https://github.com/swiftlang/swift-driver/pull/1728

**Pull request merged**